fuzz.go

```diff
@@ -503,35 +504,99 @@ type int63nPicker interface {
 	Int63n(int64) int64
 }
 
-type charRange struct {
-	first, last rune
+// UnicodeRange describes a sequential range of unicode characters.
+// Last must be numerically greater than First.
+type UnicodeRange struct {
+	First, Last rune
 }
 
+// UnicodeRanges describes an arbitrary number of sequential ranges of unicode characters.
+// To be useful, each range must have at least one character (First <= Last) and
+// there must be at least one range.
+type UnicodeRanges []UnicodeRange
+
 // choose returns a random unicode character from the given range, using the
 // given randomness source.
-func (cr charRange) choose(r int63nPicker) rune {
-	count := int64(cr.last - cr.first + 1)
-	return cr.first + rune(r.Int63n(count))
+func (ur UnicodeRange) choose(r int63nPicker) rune {
+	count := int64(ur.Last - ur.First + 1)
+	return ur.First + rune(r.Int63n(count))
+}
+
+// CustomStringFuzzFunc constructs a FuzzFunc which produces random strings.
+// Each character is selected from the range ur. If there are no characters
+// in the range (cr.Last < cr.First), this will panic.
+func (ur UnicodeRange) CustomStringFuzzFunc() func(s *string, c Continue) {
+	return func(s *string, c Continue) {
+		ur.check()
+		*s = ur.randString(c.Rand)
+	}
+}
+
+// check is a function that used to check whether the first of ur(UnicodeRange)
+// is greater than the last one.
+func (ur UnicodeRange) check() {
+	if ur.Last < ur.First {
+		panic("last encoding must be greater than the first one.")
+	}
+}
+
+// randString of UnicodeRange makes a random string up to 20 characters long.
+// Each character is selected form ur(UnicodeRange).
+func (ur UnicodeRange) randString(r *rand.Rand) string {
+	n := r.Intn(20)
+	sb := strings.Builder{}
+	sb.Grow(n)
+	for i := 0; i < n; i++ {
+		sb.WriteRune(ur.choose(r))
+	}
+	return sb.String()
 }
 
-var unicodeRanges = []charRange{
+// defaultUnicodeRanges sets a default unicode range when user do not set
+// CustomStringFuzzFunc() but wants fuzz string.
+var defaultUnicodeRanges = UnicodeRanges{
 	{' ', '~'},           // ASCII characters
 	{'\u00a0', '\u02af'}, // Multi-byte encoded characters
 	{'\u4e00', '\u9fff'}, // Common CJK (even longer encodings)
 }
 
-// randString makes a random string up to 20 characters long. The returned string
-// may include a variety of (valid) UTF-8 encodings.
-func randString(r *rand.Rand) string {
+// CustomStringFuzzFunc constructs a FuzzFunc which produces random strings.
+// Each character is selected from one of the ranges of ur(UnicodeRanges).
+// Each range has an equal probability of being chosen. If there are no ranges,
+// or a selected range has no characters (.Last < .First), this will panic.
+func (ur UnicodeRanges) CustomStringFuzzFunc() func(s *string, c Continue) {
+	return func(s *string, c Continue) {
+		// Check unicode ranges slice is empty.
+		if len(ur) == 0 {
+			panic("Unicode ranges is empty.")
+		}
+		// if not empty, each range should be checked.
+		for i := range ur {
```

Comment:
Can we move both of these checks outside of this inner function, so they're executed once and not on every invocation? Also this will move the panic message closer to where the programmer has made the error.

(Note that the panics are OK only because this is a testing library! Ordinarily, we'd want a constructor that can return an error instead, with an optional Must() function to turn that into a panic if needed. But I think that is overkill here.)

Reply:
I agree with what you said above. But I don't understand the use of the `Must` function here.

Does it mean that adding `Must` function in `UnicodeRanges` or `Must` function in `UnicodeRange`?

In Earlier time when I make this PR, I thought about making the `First` and `Last` of `UnicodeRange` as private variables.Then provide the constructor `NewUnicodeRange` to return a pointer of `UnicodeRange`.
For Example:

```go
type UnicodeRange struct {
	first, last rune
}

func NewUnicodeRange(first, last rune) (*UnicodeRange, error) {
	if last < first {
		return nil, fmt.Errorf("last encoding must be greater than the first one.")
	}

	ur := &UnicodeRange{
		first: first,
		last:  last,
	}
	return ur, nil
}
```

Do you have any suggestion about this constructor?

Reply:
If the constructor of UnicodeRange is added, what bothers me is, what should the constructor of UnicodeRange look like?

emmmm...

I thought of two methods.
One is: Constructor `NewUnicodeRanges` just return a pointer of `UnicodeRanges`, which is empty slice. Then supply 
function `Add` of `UnicodeRanges` to add `UnicodeRange` into slice. But this method requires the programmer to first call `NewUnicodeRange` to get a pointer of `UnicodeRange`
```go 
type UnicodeRanges []UnicodeRange
func (ur UnicodeRanges) Add(unicodeRange *UnicodeRange) error {
	if unicodeRange.last < unicodeRange.first {
		return fmt.Errorf("Unicode range: %v must be greater than %v", unicodeRange.Last, unicodeRange.First)
	}
	
	ur = append(ur, *unicodeRange)
	return nil 
}
```

Another is: The constructor of `NewUnicodeRanges` provides multi-parameter input of `UnicodeRange`, each `UnicodeRange` will be checked. If the conditions are not met, an error will be returned immediately. If met, the `UnicodeRange` will be added into `UnicodeRanges`, finally return the pointer of `UnicodeRanges`. 
**The first and last must be public variables**
```go
func NewUnicodeRanges(unicodeRanges ...UnicodeRange) (*UnicodeRanges, error) {
	if len(unicodeRanges) == 0 {
		return nil, fmt.Errorf("Unicode ranges is empty.")
	}

	ur := make(UnicodeRanges, 0, len(unicodeRanges))
	for i := range unicodeRanges {
		if unicodeRanges[i].Last < unicodeRanges[i].First {
			return nil, fmt.Errorf("Range: %v is bigger than %v",
				unicodeRanges[i].First, unicodeRanges[i].Last)
		}
		ur[i] = unicodeRanges[i]
	}

	return &ur, nil
}
```

Reply:
Sorry for not being clear-- I wasn't asking for a constructor here, just to move the check outside of the inner function, like this:

```go
func (ur UnicodeRanges) CustomStringFuzzFunc() func(s *string, c Continue) {
	// Check unicode ranges slice is empty.
	if len(ur) == 0 {
		panic("Unicode ranges is empty.")
	}
	// if not empty, each range should be checked.
	for i := range ur {
		ur[i].check()
	}
	return func(s *string, c Continue) {
		*s = ur.randString(c.Rand)
	}
}
```

I don't think this is so complicated that it needs a constructor. The "Must" pattern I was referring to is [here](https://golang.org/pkg/text/template/#Must), but I think that's more machinery than this code needs and we should keep it simple. :)

Reply:
You are so nice! ^ - ^

It is my first time to see the usage of `Must` function. I learned it !
Yes, I agree with you. 
Using `Must` function is more machinery. But in this test library, we need to add public `Must` function.
The programer call `Must` function, they should call the package `fuzz` first, and the `check` function should be open as public function, so that they call will like this:
```go
unicodeRange := fuzz.UnicodeRange{ 'a', 'z'}
fuzz.Must(unicodeRange.Check())
// ...
```  
The usage of `UnicodeRange` will be complicated.

I have moved the `check` function outside the inner function. 
Please @lavalamp review again, thanks~